serializers/src/main/java/io/pravega/schemaregistry/codec/Codecs.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.codec;
+
+import com.fasterxml.jackson.databind.util.ByteBufferBackedInputStream;
+import io.pravega.schemaregistry.contract.data.CodecType;
+import lombok.Getter;
+import org.apache.commons.io.IOUtils;
+import org.xerial.snappy.Snappy;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * Utility class for creating codecs for none, snappy or gzip. 
+ */
+public enum Codecs {
+    None(Constants.NOOP),
+    GzipCompressor(Constants.GZIP_CODEC), 
+    SnappyCompressor(Constants.SNAPPY_CODEC);
+
+    @Getter
+    private final Codec codec;
+    
+    Codecs(Codec codec) {
+        this.codec = codec;  
+    }
+
+    private static class Noop implements Codec {
+        private static final CodecType CODEC_TYPE_NONE = new CodecType(Constants.NONE);
+
+        @Override
+        public String getName() {
+            return CODEC_TYPE_NONE.getName();
+        }
+
+        @Override
+        public CodecType getCodecType() {
+            return CODEC_TYPE_NONE;
+        }
+
+        @Override
+        public void encode(ByteBuffer data, ByteArrayOutputStream bos) {
+        }
+
+        @Override
+        public ByteBuffer decode(ByteBuffer data, Map<String, String> codecProperties) {
+            return data;
+        }
+    }
+
+    private static class GZipCodec implements Codec {
+        private static final CodecType CODEC_TYPE_GZIP = new CodecType(Constants.APPLICATION_X_GZIP);
+        @Override
+        public String getName() {
+            return CODEC_TYPE_GZIP.getName();
+        }
+
+        @Override
+        public CodecType getCodecType() {
+            return CODEC_TYPE_GZIP;
+        }
+
+        @Override
+        public void encode(ByteBuffer data, ByteArrayOutputStream bos) throws IOException {
+            try (GZIPOutputStream gzipOS = new GZIPOutputStream(bos)) {
```

Comment:
> This is going to call close on the output stream. I am not sure we want that. Should these codecs be "stackable"? 
The codecs are not stackable. 
However, why the `close` is inconsequential here is for two reasons:
1. Encode is the last operation we do on the outputstream. 
2. we use ByteArrayOutputStream where close method is a no-op. 

So calling close on bos is inconsequential and the Serializer does not impose any restriction. However, just from the sanctity of contract perspective, i will update the javadoc to illustrate explicitly that it is inconsequential if users close the outputstream.

Reply:
Does this me we are precluding support for base64? 
for pravega, maybe it never makes sense, because we do the decode before handing the data to user code (assuming both ends use this client). But is this intended to be used in any other contexts?

Reply:
> Does this me we are precluding support for base64?
No. 
theoretically a user could create a codec for encoding and decoding with base64.

> But is this intended to be used in any other contexts?
these serializers are specific to pravega. 

Service currently generates encoding id for a single codec + schema pair. 
If users want to stack multiple encodings, presently they would create a new codec, register its name, and then in the implementation for that codec they would perform multiple encodings. 

for example:
```
compressAndBase64Encoder ==>
void encode(bytes, outputstream) {
   b1 = compress(bytes)
   b2 = base64encode(b1)
   outputstream.write(b2)
}

bytes decode(bytes) {
   b1 = base64decode(bytes)
   b2 = uncompress(b1)
   return b2
}
```

If we want we could change this in future to allow for an ordered list of "codecs" for generating an encoding id. Presently that is not the case though. 
I have created an issue to explore the possibilities https://github.com/pravega/schema-registry/issues/56